lib/SILOptimizer/Mandatory/TFConstExpr.cpp

```diff
@@ -996,146 +1010,66 @@ SymbolicValue ConstExprFunctionState::getConstantValue(SILValue value) {
-      assert(elementNo < tuple->getNumElements() && "invalid index");
-      eltType = tuple->getElement(elementNo).getType();
-    } else {
-      return true;
-    }
-  }
-
-  if (writeOnlyOnce &&
-      oldElts[elementNo].getKind() != SymbolicValue::UninitMemory) {
-    // Cannot overwrite an existing constant.
-    return true;
-  }
-
-  // Update the indexed element of the aggregate.
-  SmallVector<SymbolicValue, 4> newElts(oldElts.begin(), oldElts.end());
-  if (updateIndexedElement(newElts[elementNo], indices.drop_front(), scalar,
-                           eltType, writeOnlyOnce, allocator))
-    return true;
-
-  if (aggregate.getKind() == SymbolicValue::Aggregate)
-    aggregate = SymbolicValue::getAggregate(newElts, allocator);
-  else
-    aggregate = SymbolicValue::getArray(newElts, eltType->getCanonicalType(),
-                                        allocator);
-  return false;
-}
-
-/// Find the initializer (single writer) of `addr` among it users,
-/// const-evaluate it and store the result into a memory object, and return the
-/// address of that memory object.
+/// This is a helper function for `getSingleWriterAddressValue`. Callers should
+/// use `getSingleWriterAddressValue`.
 ///
-/// Some use cases are:
-/// 1. When analyzing the top-level code involved in a constant expression, we
-/// can end up demanding values that are returned by address.  Handle this by
-/// finding the temporary stack value (an alloc_stack inst), and calling this
-/// method on it.
-/// 2. When const-evaluating an array via decodeAllocUninitializedArray(),
-/// do that by const-evaluating the writers of individual array elements.
+/// If `addr` has no writing uses, returns {false, None} and has no side
+/// effects.
 ///
-///  There are a few forms of writers, such as:
-/// - store %3 to %4 ...
-/// - %8 = pointer_to_address %7 : $Builtin.RawPointer to [strict] $*Int32
-/// - %14 = index_addr %9 : $*Int32, %13 : $Builtin.Word
-/// - %180 = tuple_element_addr %179 : $*(Int32, Int32, Int32, Int32), 3
+/// If the following conditions hold:
+///   * `addr` points at uninitialized memory;
+///   * there are store(s) to `addr` that, taken together, set the memory
+///     exactly once (e.g. a single "store" to `addr` OR multiple "store"s to
+///     different "tuple_element_addr"s of `addr`); and
+///   * the stores' value(s) can be const-evaluated;
+/// Then: initializes the memory at `addr` and returns {true, None}.
 ///
-///  Note unlike getConstAddrAndLoadResult(), this method does *not*
-///  const-evaluate the input `addr` by evaluating its operand first, such as %7
-///  above. Instead, it finds a user of %8 who is the initializer, and uses that
-///  to set the const value for %7. In other words, this method propagates const
-///  info from result to operand (e.g. from %8 to %7), while
-///  getConstAddrAndLoadResult() propagates const info from operand to result.
+/// Otherwise, sets the memory at `addr` to an unspecified value and returns
```

Comment:
in this case, do we need to set `addr` to unknown, so that any const-eval later over addr will be guaranteed to fail?

if we intend for the caller (the wrapper) to do that, then consider replacing "sets the memory at `addr` to an unspecified value" with "the memory at `addr` is undefined". in that case it might be good to also mention the wrapper will set it to unknown -- the code in (const eval) caller depends on that invariant being enforced / restored.

Reply:
Setting the memory at `addr` to unknown shouldn't be necessary for correctness, because the failure will propagate all the way up to the wrapper and then the wrapper will set the entire memory, which includes the memory at `addr` to unknown.

However, setting the memory at `addr` to unknown makes the function's contract easier to understand so I'll change it to do that.